### PR TITLE
Backend registration verification code

### PR DESCRIPTION
**Pull request by mo-ayman**
Added

- register verification api end point
- register verification service logic

Database

- database separate verified user from unverified
- database add verification codes table
- database separate revision token from user entity (Normalization)

### CODE REVIEW

File: Student-Unity/src/main/java/csed/swe/studentunity/DAO/UnverifiedUserRepo.java [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package csed.swe.studentunity.DAO;
+
+import csed.swe.studentunity.model.UnverifiedUser;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+import java.util.Optional;
+
+
+public interface UnverifiedUserRepo extends JpaRepository<UnverifiedUser, Long> {
+    Optional<UnverifiedUser> findUnverifiedUserByEmail(String email);
```
Comment (Ahmedelsa3eed) [inline Student-Unity/src/main/java/csed/swe/studentunity/DAO/UnverifiedUserRepo.java:10]: we made a primary key to the user instead of it's email, you should reference the user now by it's **id** not email for consistency.

File: Student-Unity/src/main/java/csed/swe/studentunity/DAO/VerificationCodeRepo.java [hunk 2]
```diff
@@ -0,0 +1,10 @@
+package csed.swe.studentunity.DAO;
+
+import csed.swe.studentunity.model.VerificationCode;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+import java.util.Optional;
+
+public interface VerificationCodeRepo extends JpaRepository<VerificationCode, Long> {
+    Optional<VerificationCode> findVerificationCodeByEmail(String email);
```
Comment (Ahmedelsa3eed) [inline Student-Unity/src/main/java/csed/swe/studentunity/DAO/VerificationCodeRepo.java:9]: **ById**

File: Student-Unity/src/main/java/csed/swe/studentunity/Logic/RegistrationService.java [hunk 3]
```diff
@@ -1,37 +1,61 @@
...
 import csed.swe.studentunity.model.User;
+import csed.swe.studentunity.model.VerificationRequest;
 import lombok.AllArgsConstructor;
 import org.springframework.stereotype.Service;
 
+import java.util.Optional;
+
 @AllArgsConstructor
 @Service
 //@Transactional
```
Comment (Ahmedelsa3eed) [inline Student-Unity/src/main/java/csed/swe/studentunity/Logic/RegistrationService.java:14]: if you don't need this commented code, please **delete** it

File: Student-Unity/src/main/java/csed/swe/studentunity/Logic/RegistrationService.java [hunk 4]
```diff
@@ -1,37 +1,61 @@
...
+    private final UnverifiedUserService unverifiedUserService;
+    private final EmailService emailService;
 
     public String addUser(RegistrationRequest request) {
         // check if email is already in use, if not: add user to database and send verification email
-        if(userService.getUser(request.getEmail()).isEmpty()) {
-            User newUser = new User(request.getEmail(), request.getStudentId(), request.getFirstName(),
-                    request.getLastName(), request.getPassword(), "student", "");
-            userService.addUser(newUser);
+        if(unverifiedUserService.getUnverifiedUser(request.getEmail()).isEmpty()
```
Comment (Ahmedelsa3eed) [inline Student-Unity/src/main/java/csed/swe/studentunity/Logic/RegistrationService.java:23]: the service name **unverifiedUserService** is misleading. 
why it's prefixed with ``un``

File: Student-Unity/src/main/java/csed/swe/studentunity/Logic/RegistrationService.java [hunk 5]
```diff
@@ -1,37 +1,61 @@
...
+            return "User verification failed";
+        }
+
     }
 
-    public String generateCode() {
-        // milestone 2
-        return "works";
+    private User convertUnverifiedUserToUser(UnverifiedUser unverifiedUser) {
+        return new User(unverifiedUser.getEmail(), unverifiedUser.getStudentId(), unverifiedUser.getFirstName(),
```
Comment (Ahmedelsa3eed) [inline Student-Unity/src/main/java/csed/swe/studentunity/Logic/RegistrationService.java:55]: we could use builder design pattern in user class

File: Student-Unity/src/main/java/csed/swe/studentunity/model/RevisionNotification.java [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+import lombok.ToString;
+
+@Getter
+@Setter
+@ToString
+@Entity(name = "RevisionNotification")
+public class RevisionNotification {
+    @Id
+    @Column(name = "user_id")
+    private Long id;
```
Comment (Ahmedelsa3eed) [inline Student-Unity/src/main/java/csed/swe/studentunity/model/RevisionNotification.java:15]: what id is that ???????? 
tokenId or userId
if you mead userId, then the ``User user`` property is enough as it will refrence the user as a foreign key

File: Student-Unity/src/main/java/csed/swe/studentunity/model/User.java [hunk 7]
```diff
@@ -19,7 +19,8 @@ public class User implements Serializable {
 
     @Id
     @GeneratedValue(strategy = GenerationType.SEQUENCE)
-    @Column(name = "id", nullable = false, columnDefinition = "INTEGER")
+    @Column(name = "id", updatable = false)
```
Comment (Ahmedelsa3eed) [inline Student-Unity/src/main/java/csed/swe/studentunity/model/User.java:22]: why did you remove ``columnDefinition = "INTEGER"``

**Review by Ahmedelsa3eed — CHANGES_REQUESTED** [review]

File: Student-Unity/src/main/java/csed/swe/studentunity/DAO/VerificationCodeRepo.java (hunk 2 above)
Comment (mo-ayman) [reply Student-Unity/src/main/java/csed/swe/studentunity/DAO/VerificationCodeRepo.java:9]: If you notice this table involves the verification codes of forget password and code of registered unverified users
and there are separate tables for user and unverified user so i can't use id of user and unverified user here as it may be equal each other as the id is generated sequentially so the only thing to uniquely identify the row in this table is by the email

File: Student-Unity/src/main/java/csed/swe/studentunity/Logic/RegistrationService.java (hunk 4 above)
Comment (mo-ayman) [reply Student-Unity/src/main/java/csed/swe/studentunity/Logic/RegistrationService.java:23]: To keep naming convention as this class work with the table of unverified users in database

Screenshot from 2022-12-16 22-51-19

File: Student-Unity/src/main/java/csed/swe/studentunity/Logic/RegistrationService.java [hunk 8]
```diff
@@ -1,37 +1,61 @@
...
-                    request.getLastName(), request.getPassword(), "student", "");
-            userService.addUser(newUser);
+        if(unverifiedUserService.getUnverifiedUser(request.getEmail()).isEmpty()
+        && userService.getUser(request.getEmail()).isEmpty()) {
+            UnverifiedUser newUser = new UnverifiedUser(request.getEmail(), request.getStudentId(), request.getFirstName(),
+                    request.getLastName(), request.getPassword(), "student");
+            String code = unverifiedUserService.addUnverifiedUser(newUser);
+            emailService.send("You have registered in Student Unity \n" +
+                    "verification code is: " + code, "verification code", request.getEmail());
             return "User registered successfully";
```
Comment (AhmedElemary57) [inline Student-Unity/src/main/java/csed/swe/studentunity/Logic/RegistrationService.java:30]: I think that it is better to return enums instead of strings as checking for string in the API layer will be not good But it is fine we can do it later

**Review by AhmedElemary57 — DISMISSED** [review]

File: Student-Unity/src/main/java/csed/swe/studentunity/model/RevisionNotification.java (hunk 6 above)
Comment (mo-ayman) [reply Student-Unity/src/main/java/csed/swe/studentunity/model/RevisionNotification.java:15]: It's a foreign key refers to user id 
user property doesn't mean that this will save user as a field in this table it's a syntax of how you make foreign key in spring data jpa to map user_id field to id in user table

As you see in the screen shot down below revision notification table doesn't have user property
Screenshot from 2022-12-16 22-59-26

File: Student-Unity/src/main/java/csed/swe/studentunity/DAO/VerificationCodeRepo.java (hunk 2 above)
Comment (Ahmedelsa3eed) [reply Student-Unity/src/main/java/csed/swe/studentunity/DAO/VerificationCodeRepo.java:9]: ok no problem

File: Student-Unity/src/main/java/csed/swe/studentunity/Logic/RegistrationService.java (hunk 4 above)
Comment (Ahmedelsa3eed) [reply Student-Unity/src/main/java/csed/swe/studentunity/Logic/RegistrationService.java:23]: ok good job

File: Student-Unity/src/main/java/csed/swe/studentunity/DAO/UnverifiedUserRepo.java (hunk 1 above)
Comment (mo-ayman) [reply Student-Unity/src/main/java/csed/swe/studentunity/DAO/UnverifiedUserRepo.java:10]: But unverified user table is for users who didn't verify submit a code sent to his email.
so when a user gets the verification code from his email and he submit this code the only thing that i have to identify this user in the table is using his email.
we use user id when a user has already logged in as i save his id in session id. 
but in this situation this is for unverified user.

File: Student-Unity/src/main/java/csed/swe/studentunity/model/User.java (hunk 7 above)
Comment (mo-ayman) [reply Student-Unity/src/main/java/csed/swe/studentunity/model/User.java:22]: Actually i didn't remove it for intended reason someone else have added it and i didn't make a pull before i pushed my changes so it was removed.
I have no problem to keep it.

File: Student-Unity/src/main/java/csed/swe/studentunity/Logic/RegistrationService.java (hunk 8 above)
Comment (mo-ayman) [reply Student-Unity/src/main/java/csed/swe/studentunity/Logic/RegistrationService.java:29]: You are right. I intend to do that.
But I'm waiting for our next scrum to specify what should be returned.

File: Student-Unity/src/main/java/csed/swe/studentunity/Logic/RegistrationService.java (hunk 5 above)
Comment (mo-ayman) [reply Student-Unity/src/main/java/csed/swe/studentunity/Logic/RegistrationService.java:53]: Good idea.
I will do that.

**Review by Ahmedelsa3eed — APPROVED** [review]

**Review by AhmedElemary57 — APPROVED** [review]
